### PR TITLE
Tree restore tech debt

### PR DESCRIPTION
**Pull request by BaldyAsh**
#157

### CODE REVIEW

File: core/data_restore/src/accounts_state.rs [hunk 1]
```diff
@@ -189,136 +189,239 @@ impl FranklinAccountsState {
...
-        println!("updates 5 {:?} \n", updates5);
-        println!("root hash 5 {:?} \n", tree.root_hash());
-        println!("accounts 5 {:?} \n", tree.get_accounts());
+        tree.update_accounts_states_from_ops_block(&block5)
+            .expect("Cant update state from block 5");
 
         assert_eq!(
             tree.root_hash().to_string(),
-            "Fr(0x0f220069602ed8f8c4557fdde71baf5220bbf237790adf67f49280b84588acf2)".to_string()
+            "Fr(0x06210ec05cbdd8a974447f23aa9a2e2a6c6fe500620bd4ffc0d9752d1034a062)".to_string()
```
Comment (dvush) [inline core/data_restore/src/accounts_state.rs]: Please no. No hardcoded values like that. What if someone would like to change something in our parameters or elsewhere?

File: core/data_restore/src/accounts_state.rs [hunk 2]
```diff
@@ -189,136 +189,239 @@ impl FranklinAccountsState {
...
         );
     }
 
     #[test]
-    // TODO: Fix test of tree restore
-    #[ignore]
     fn test_tree_inconsistent_update() {
```
Comment (dvush) [inline core/data_restore/src/accounts_state.rs]: What outcome of this test is expected? If it is "inconsistent update" then why this test should pass?

File: core/data_restore/src/accounts_state.rs [hunk 3]
```diff
@@ -189,136 +189,239 @@ impl FranklinAccountsState {
...
-        println!("root hash 1 {:?} \n", tree.root_hash());
-        println!("accounts 1 {:?} \n", tree.get_accounts());
+        let updates = tree
+            .update_accounts_states_from_ops_block(&block)
+            .expect("Cant update state from block");
 
         assert_eq!(
             tree.root_hash().to_string(),
-            "Fr(0x0f220069602ed8f8c4557fdde71baf5220bbf237790adf67f49280b84588acf2)".to_string()
+            "Fr(0x06210ec05cbdd8a974447f23aa9a2e2a6c6fe500620bd4ffc0d9752d1034a062)".to_string()
```
Comment (dvush) [inline core/data_restore/src/accounts_state.rs]: same

File: core/models/src/node/operations.rs [hunk 4]
```diff
@@ -45,15 +41,37 @@ impl DepositOp {
     }
 
     pub fn from_bytes(bytes: &[u8]) -> Option<Self> {
-        if bytes.len() != Self::OP_LENGTH {
+        if bytes.len() != Self::CHUNKS * 8 {
             return None;
         }
-        let pre_length = 0;
+        let account_id_pre_length = 1;
+        let token_id_pre_length = account_id_pre_length + ACCOUNT_ID_BIT_WIDTH / 8;
```
Comment (dvush) [inline core/models/src/node/operations.rs]: `pre_length` -> `offset`

File: core/models/src/node/operations.rs [hunk 5]
```diff
@@ -225,15 +333,23 @@ impl CloseOp {
     }
 
     pub fn from_bytes(bytes: &[u8]) -> Option<Self> {
```
Comment (dvush) [inline core/models/src/node/operations.rs]: use `Result<.., failure::Error>` for error messages.
There are also a lot of useful macro from failure crate - `bail`, `ensure`

File: core/models/src/node/operations.rs [hunk 6]
```diff
@@ -335,20 +483,8 @@ impl FranklinOp {
...
-            CloseOp::OP_CODE => Some(CloseOp::CHUNKS),
-            TransferOp::OP_CODE => Some(TransferOp::CHUNKS),
-            FullExitOp::OP_CODE => Some(FullExitOp::CHUNKS),
-            _ => None,
-        }
-    }
-
-    pub fn from_bytes(op_type: u8, bytes: &[u8]) -> Option<Self> {
+    pub fn from_bytes(bytes: &[u8]) -> Option<Self> {
+        let op_type: u8 = bytes[0];
```
Comment (dvush) [inline core/models/src/node/operations.rs]: Unnecessary panic. `[0]` is dangerous

File: core/models/src/node/operations.rs [hunk 7]
```diff
@@ -268,31 +383,64 @@ impl FullExitOp {
     }
 
     pub fn from_bytes(bytes: &[u8]) -> Option<Self> {
```
Comment (dvush) [inline core/models/src/node/operations.rs]: `from_bytes` -> `from_public_data`

File: core/data_restore/src/storage_interactor.rs [hunk 8]
```diff
@@ -27,15 +27,16 @@ pub fn update_tree_state(
...
+) -> Result<(), failure::Error> {
+    let storage = connection_pool.access_storage().map_err(|e| {
+        format_err!(
+            "Db connection failed for data restore save tree state: {}",
+            e.to_string()
         )
     })?;
     storage
         .update_tree_state(block_number, &account_updates)
-        .map_err(|_| DataRestoreError::Storage("cant save tree state".to_string()))?;
```
Comment (dvush) [inline core/data_restore/src/storage_interactor.rs]: there was no need to remove `DataRestoreError` - it was not bad at all.
you could use `#[derive(Fail)]` see example https://github.com/rust-lang-nursery/failure

File: core/models/src/node/operations.rs [hunk 9]
```diff
@@ -40,26 +40,25 @@ impl DepositOp {
         data
     }
 
-    pub fn from_bytes(bytes: &[u8]) -> Option<Self> {
+    pub fn from_public_data(bytes: &[u8]) -> Option<Self> {
```
Comment (dvush) [inline core/models/src/node/operations.rs]: I was talking about using Result/failure here. 
This way you could know why transform from pubdata failed.

`.ok()?` and `return None` is non informative.

File: core/data_restore/src/storage_interactor.rs (hunk 8 above)
Comment (BaldyAsh) [reply core/data_restore/src/storage_interactor.rs]: Agree, but I have not seen other error enums anywhere else in the code. And now I see it as an "extra code". Think we can use failure::Error everywhere

File: core/models/src/node/operations.rs (hunk 9 above)
Comment (BaldyAsh) [reply core/models/src/node/operations.rs]: Done just a few moments after you've written this

**Pull request merged**